### PR TITLE
fix(mo.ui.table): implement row count limit on column summaries; various bug fixes

### LINKED ISSUE #1979: Automatic column summaries of the table component should have a upper row count limit
### Describe the bug

I'm using marimo to work on large datasets with 100 million rows, and found the "run cell" button unresponsive sometimes. After some investigation, I found during the unresponsive time, the kernel is busy running the `get_column_summaries` function, which is issued by the table component in the frontend.

Calculating summaries for 100m rows will take huge amounts of time, and the unresponsive behavior will cause confusions to users. I think marimo should impose a hard limit on column summary feature. Maybe 1 million is a good number?

Or maybe we can try to run the column summary in a background thread. But this will be a huge change I think, and thread safety issue will be hard to deal with.

### Environment

```
{
  "marimo": "0.7.14",
  "OS": "Linux",
  "OS Version": "5.15.0-52-generic",
  "Processor": "x86_64",
  "Python Version": "3.10.12",
  "Binaries": {
    "Browser": "--",
    "Node": "--"
  },
  "Requirements": {
    "click": "8.1.7",
    "importlib-resources": "missing",
    "jedi": "0.19.1",
    "markdown": "3.6",
    "pymdown-extensions": "10.9",
    "pygments": "2.18.0",
    "tomlkit": "0.13.0",
    "uvicorn": "0.30.3",
    "starlette": "0.38.2",
    "websockets": "12.0",
    "typing-extensions": "4.12.2",
    "ruff": "0.5.5"
  }
}
```

### Code to reproduce

_No response_

### LINKED ISSUE #2016: mo.ui.table()'s column summaries are not updated after changing search/filter criteria
### Describe the bug

When I'm trying to fix #1979, I see the implementation of `get_column_summaries` uses `self._filtered_manager`:

https://github.com/marimo-team/marimo/blob/3969f95cde6146780d1fc3e856965d5675cd152f/marimo/_plugins/ui/_impl/table.py#L340-L344

So I thought a smooth user experience after applying row limit on column summaries would be: user shows a large table -> marimo complains about too many rows -> user applies filter/search -> marimo happily displays column summaries.

But after I implemented the backend logic, I found the column summaries are not updated after changing the filter (no new `get_column_summaries` function call in DevTools' network tab). I think it might be related to the `columnSummaries` are not put inside a `useEffect` in the following code (I'm completely new to React so correct me if I'm wrong):

https://github.com/marimo-team/marimo/blob/3969f95cde6146780d1fc3e856965d5675cd152f/frontend/src/plugins/impl/DataTablePlugin.tsx#L252-L265

I'm not sure whether it is a bug or working as intended. Posting here so maintainers are aware of it.

### Environment

```
{
  "marimo": "0.7.19",
  "OS": "Linux",
  "OS Version": "5.15.0-52-generic",
  "Processor": "x86_64",
  "Python Version": "3.10.12",
  "Binaries": {
    "Browser": "--",
    "Node": "--"
  },
  "Requirements": {
    "click": "8.1.7",
    "importlib-resources": "missing",
    "jedi": "0.19.1",
    "markdown": "3.6",
    "pymdown-extensions": "10.9",
    "pygments": "2.18.0",
    "tomlkit": "0.13.0",
    "uvicorn": "0.30.3",
    "starlette": "0.38.2",
    "websockets": "12.0",
    "typing-extensions": "4.12.2",
    "ruff": "0.5.5"
  }
}
```

### Code to reproduce

_No response_

### PR DESCRIPTION
**Pull request by mutongx**
## 📝 Summary

1. Add a hard-coded row count upper limit on table's column summaries feature.
2. Update column summaries after changing search/filter criteria
3. Fix the inconsistency of table filtering and row count limiting
4. Some code reorganization

Fixes #1979.
Fixes #2016.

## 🔍 Description of Changes

- Add a `is_disabled` field to `ColumnSummaries` so it can indicate itself being disabled
    - I tried to use exceptions, but I found Function objects only propagates human-readable error messages. So it's unfeasible.
- Add a hard-coded 1,000,000 row count limit to the column summaries feature
    - marimo already imposes a column count limit, so I think limiting only rows here should be OK for most situations
- Make column summaries react to filter/search criteria changes, so the summaries gets updated after such operations
- Separate the `table._filtered_manager` to `table._searched_manager` and `table._limited_manager`
    - The `_searched_manager` holds the full data after search (query, filter and sort), and the `_limited_manager` holds the first 20000 rows of `_searched_manager`
- Reorganized the code in `table.__init__()` for better clarity

## 📋 Checklist

- [x] I have read the [contributor guidelines](https://github.com/marimo-team/marimo/blob/main/CONTRIBUTING.md).
- [ ] (Not applicable) For large changes, or changes that affect the public API: this change was discussed or approved through an issue, on [Discord](https://discord.gg/JE7nhX6mD8), or the community [discussions](https://github.com/marimo-team/marimo/discussions) (Please provide a link if applicable).
- [x] I have added tests for the changes made.
- [x] I have run the code and verified that it works as expected.

## 📜 Reviewers

@akshayka OR @mscolnick

### CODE REVIEW

**Comment by akshayka** [comment]
Looking at this, I'm reminded that getting the number rows can be expensive for some tables; @mscolnick, is that a concern here, or was that unrelated?

File: marimo/_plugins/ui/_impl/table.py
Comment (akshayka) [inline marimo/_plugins/ui/_impl/table.py:71]: Doesn't `summaries` being the empty list imply that column summaries are disabled? In which case we could remove the `is_disabled` parameter.

**Review by akshayka — COMMENTED** [review]
Thanks for the PR! One drive-by comment on the Python code. I'll let @mscolnick review more thoroughly, he is the owner of tables.

File: frontend/src/plugins/impl/DataTablePlugin.tsx
Comment (mscolnick) [inline frontend/src/plugins/impl/DataTablePlugin.tsx:271]: you are right - it looks like we were missing these deps so they were not re-rendering on filter changes. thanks for fixing

File: frontend/src/plugins/impl/DataTablePlugin.tsx
Comment (mscolnick) [inline frontend/src/plugins/impl/DataTablePlugin.tsx:198]: i dont think we need to keep track of this state. this is the same as what is inside `useAsyncData`

File: marimo/_plugins/ui/_impl/table.py
Comment (mscolnick) [reply marimo/_plugins/ui/_impl/table.py:71]: it is helpful to know if there are no summaries vs it actually being disabled. 

it could have no summaries if they failed for some reason or summaries are not supported in the table type

File: marimo/_plugins/ui/_impl/tables/table_manager.py
Comment (mscolnick) [inline marimo/_plugins/ui/_impl/tables/table_manager.py:29]: could we add a comment here or link to the ticket, just so no one changes this without investigating.

File: frontend/src/plugins/impl/DataTablePlugin.tsx
Comment (mscolnick) [inline frontend/src/plugins/impl/DataTablePlugin.tsx:266]: i think it is fine to show the previous summaries while the new ones are loading, otherwise it might flicker a bunch

File: marimo/_plugins/ui/_impl/table.py
Comment (mutongx) [reply marimo/_plugins/ui/_impl/table.py:71]: Thanks @mscolnick for the explanation!

A better way is to implement a generic error handling for Function objects, so we can pass structural error information from the backend. Actually I even wanted to add the upper limit number to the `ColumnSummaries` object so there will be no hardcoded text in frontend (I didn't because it's even uglier than adding `is_disabled`).

File: frontend/src/plugins/impl/DataTablePlugin.tsx
Comment (mutongx) [reply frontend/src/plugins/impl/DataTablePlugin.tsx]: Should be fixed. I remember my first try of simply adding the dependency to `useAsyncData` failed but this time it worked. Not sure why though.

File: marimo/_plugins/ui/_impl/tables/table_manager.py
Comment (mutongx) [reply marimo/_plugins/ui/_impl/tables/table_manager.py:32]: Comments are added. I also added a comment for `DEFAULT_ROW_LIMIT` and `DEFAULT_COL_LIMIT`, and please take a look. Thanks.

File: frontend/src/plugins/impl/DataTablePlugin.tsx
Comment (mutongx) [reply frontend/src/plugins/impl/DataTablePlugin.tsx]: Fixed. Thanks!

File: marimo/_plugins/ui/_impl/table.py
Comment (mscolnick) [reply marimo/_plugins/ui/_impl/table.py:71]: I agree structured errors would be great. we don't have great e2e type-safety yet for structured errors, so its fine to put the details inside the 200 response. 

i think you can add the upper limit as well. that seems like a good idea in case its configurable later.

Comment (mscolnick) [reply marimo/_plugins/ui/_impl/table.py:71]: keeping in sync via comments is also totally great, this looks great!

**Review by mscolnick — APPROVED** [review]
looks great! thanks for the contribution!

**Comment by mutongx** [comment]
@mscolnick 

Sorry but please don't merge this PR for now. I think I found some bug in the table component's `search()` function.

When I'm writing unit tests, the following test code fails the assert:

```python
def test_table_with_too_many_rows_column_summaries_disabled():
    data = {"a": list(range(TableManager.DEFAULT_SUMMARY_ROW_LIMIT + 1))}
    table = ui.table(data)
    summaries = table.get_column_summaries(EmptyArgs())
    assert summaries.is_disabled is True
```

It's because `table._filtered_manager` actually gets limited to 20000 rows in the following line, so `_filtered_manager` is expected to never reach the 1m limit.

https://github.com/marimo-team/marimo/blob/60b3e5e5c3159a7bfa48fb563cc69f038dd5ff79/marimo/_plugins/ui/_impl/table.py#L262

But `_filtered_manager` is reset to the original `self._manager` in `search()` function:

https://github.com/marimo-team/marimo/blob/60b3e5e5c3159a7bfa48fb563cc69f038dd5ff79/marimo/_plugins/ui/_impl/table.py#L359-L361

and doesn't get limited:

https://github.com/marimo-team/marimo/blob/60b3e5e5c3159a7bfa48fb563cc69f038dd5ff79/marimo/_plugins/ui/_impl/table.py#L384-L386

So there might be some inconsistency in the semantic of `_filtered_manager` here. If we expect `_filtered_manager` to be always less than 20000 rows, then this PR will be invalid. However, a column summary of only 20000 rows is not useful at all.

What do you think?

**Comment by mutongx** [comment]
This should be easy to fix though. Remove the `self._filtered_manager = self._filtered_manager.limit(row_limit)` in table constuctor, and instead only apply the 20000 row limit when sending data to the frontend. I can do this in the PR but I think I need to sort out the logic so there are no potential problems.

**Comment by mscolnick** [comment]
@mutongx , yep i agree with that. I think this might just be an existing bug/quirk. 

You could do something like this:
```diff
        has_more = total_rows == "too_many" or total_rows > row_limit
+        limited_manager = self._filtered_manager
        if has_more:
+            limited_manager = self._filtered_manager.limit(row_limit)

        # pagination defaults to True if there are more than 10 rows
        if pagination is None:
            pagination = total_rows == "too_many" or total_rows > 10

        field_types = self._manager.get_field_types()

        super().__init__(
            component_name=table._name,
            label=label,
            initial_value=[],
            args={
                "data": limited_manager.to_data(format_mapping),
```

**Comment by mscolnick** [comment]
thanks for investigating this and fixing these issues

File: tests/_plugins/ui/_impl/test_table.py
Comment (mscolnick) [inline tests/_plugins/ui/_impl/test_table.py:317]: can you add another test that when we search/filter that brings it down to less than 10, summaries is re-enabled?

File: marimo/_plugins/ui/_impl/table.py
Comment (mutongx) [inline marimo/_plugins/ui/_impl/table.py:273]: The original logic is quite complex. Simply checking for `is None` should be fine, as the `_internal_total_rows` are only used in the following situations:

1. `marimo._sql` ([link](https://github.com/marimo-team/marimo/blob/60b3e5e5c3159a7bfa48fb563cc69f038dd5ff79/marimo/_sql/sql.py#L93)), where the value will be `"too_many"` or `None`
2. `test_table.py` ([link](https://github.com/marimo-team/marimo/blob/main/tests/_plugins/ui/_impl/test_table.py)), where the value should always be non-zero

File: marimo/_plugins/ui/_impl/table.py
Comment (mutongx) [inline marimo/_plugins/ui/_impl/table.py:252]: This rewrite is to keep the code style consistent with `if _internal_total_rows is None` below.

File: tests/_plugins/ui/_impl/test_table.py
Comment (mutongx) [reply tests/_plugins/ui/_impl/test_table.py]: Added. Thanks for reminding!

**Review by mscolnick — APPROVED** [review]
thanks again for the addition!

**Pull request merged**
